File: Validate.R
# Validate.R

Validate <- function(dfTraining, dfValidation, alphas,
                     Train, Predict, Loss,
                     extra=NULL,
                     verbose=1) {
  # Perform validation to select best parameters.
  #
  # Args:
  #
  # dfTraining:  a data frame containing both input and target variables. Used
  # for training the models.
  #
  # dfValidation: a data frame containing both input and target variables. Used
  # for validation the trained models.
  #
  # alphas: a list or vector; each element is an arbitrary object;
  #   An element of alphas is passed to the train and predict functions.
  #
  # Train(dfTraining, alpha, extra): Train a model using parameters alpha.
  #   dfTraining is the value passed to Validate.
  #   alpha is one of the alphas.
  #   extra is the argument passed to Validate.
  #   Value: an arbitrary object called "trained".
  #
  # Predict(trained, dfValidation, alpha, extra): Predict using trained
  # model
  #   trained is a value returned by the Train function.
  #   dfValidation is the value passed to Validate.
  #   alpha is one of the alphas.
  #   extra is the argument passed to CrossValidate.
  #   Value: a vector of predictions. The class of each element is arbitrary.
  #
  # Loss(dfValidation, predictions, extra): Determine losses from the
  # predictions.
  #   dfValidation is the value passed to Validate.
  #   predictions is the value of the Predict function.
  #   extra is the argument passed to Validate.
  #   Value: a vector of numbers, each a loss determined from the target
  #     and prediction value.
  #
  # verbose: scalar integer, verbosity level; 
  #   0 means no printing
  #   1 means print average loss for each value of alpha
  #   2 means print trace of computation
  #
  # Value: a list
  # $bestAlpha:        the element of the alphas that had the lowest loss among
  #                    the trained models
  # $bestTrainedModel: the fitted model for the best alpha
  # $bestPredictions:  the vector of predictions for the best alpha
  # $bestLoss:         the loss for the best alpha

  # determine alpha with lowest average loss
  bestAlpha <- NULL
  bestTrainedModel <- NULL
  bestPredictions <- NULL
  bestLoss <- NULL
  minLoss <- Inf
  allLosses <- NULL
  for (alpha in alphas) {
    trainedModel <- Train(dfTraining, alpha, extra)
    predictions <- Predict(trainedModel, dfValidation, alpha, extra)
    loss <- Loss(dfValidation, predictions, alpha, extra)
    allLosses <- c(allLosses, loss)
    if (verbose >= 1) {
      cat(sprintf("Validate: loss is %f for alpha\n",
                  loss))
      print(alpha)
    }

    if (loss < minLoss) {
      bestAlpha <- alpha
      bestTrainedModel <- trainedModel
      bestPredictions <- predictions
      minLoss <- loss
    }
  }
  
  if (verbose >= 1) {
    cat(sprintf("Validate: lowest loss is %f for alpha\n",
                minLoss))
    print(bestAlpha)
  }

  list(bestAlpha=bestAlpha,
       bestTrainedModel=bestTrainedModel,
       bestPredictions=bestPredictions,
       bestLoss=minLoss,
       allLosses=allLosses)
}

File: validate.lua
-- validate.lua
-- Select model based on validation process

-- Compare models
-- ARGS
-- alphas  : seq of arbitrary objects, each identifying a model
-- train   : function(alpha) returning a model trained on training data
-- predict : function(trainedModel, alpha) return a table of predictions on
--           validation data
-- loss    : function(predictions, alpha) returning a number
--           the average loss on the predictions
-- verbose : optional number, default 0, verbose level
-- RETURNS
-- losses  : seq of losses, where losses[i] is the loss for alpha[i]
function validate(alphas,
                  train, predict, loss,
                  verbose)
   -- validate args and set defaults
   assert(type(alphas) == 'table', 'alphas not a sequence')
   assert(type(train) == 'function', 'train not a function')
   assert(type(predict) == 'function', 'predict not a function')
   assert(type(loss) == 'function', 'loss not a function')
   local verbose = verbose or 0

   local vp = makeVp(verbose, 'validate')

   -- determine loss on each model
   local losses = {}
   for i, alpha in ipairs(alphas) do
      local trainedModel = train(alpha)
      local predictions = predict(trainedModel, alpha)
      local loss = loss(predictions, alpha)
      losses[i] = loss
      vp(1, 'alpha', alpha)
      vp(1, 'loss', loss)
   end

   -- losses[i] = loss on trained model using alpha[i]
   return losses 
end
